### PR TITLE
Implement annotations feature with remark and rehype plugins

### PR DESCRIPTION
**Pull request by imballinst**
This PR implements Annotations feature with footnote syntax. This is possible with the help of:

1. `remark-plugin-annotations`: this is a new plugin that we use so that we could "replace" the node with type `footnoteDefinition` to be a normal paragraph instead. It is needed, because otherwise, the Remark plugins will try to find its "pair" (outside of code blocks) and if not found, it will be discarded. By replacing it with a normal paragraph, we preserve the contents.
2. `rehype-plugin-codeblock`: this will replace the syntax `[^1]` to be a container div with 2 children: popper div and button element. The button element will have event listeners "attached" to it from the React `ProtosaurusAnnotation` component.
3. `ProtosaurusAnnotation.tsx`: this component renders nothing, but it has `useEffect` that will create the popper instances and attach event listeners to the button and popper elements.

image

Signed-off-by: Try Ajitiono <ballinst@gmail.com>

### CODE REVIEW

**Comment by dio** [comment]
What's stopping us to use footnote annotation? https://github.blog/changelog/2021-09-30-footnotes-now-supported-in-markdown-fields/ Is it because we are inside a code block?

**Comment by imballinst** [comment]
> What's stopping us to use footnote annotation? https://github.blog/changelog/2021-09-30-footnotes-now-supported-in-markdown-fields/ Is it because we are inside a code block?

Yes, it's because inside code blocks everything is rendered as-is (a plain text). See also: https://docusaurus.io/docs/markdown-features/code-blocks#using-jsx-markup

> Code block in Markdown always preserves its content as plain text

More than that, I think we want a less intrusive interaction? Basically "copying" what MkDocs Material has: https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#adding-annotations. If we click on footnote, then our screen scroll possibly changes. Sure, we can use the "return" button, but why do that if we can keep the content in a popper instead?

Alternatively, perhaps we can implement an opt-in -- whether to use the "popper" approach or use the "footnote" approach. What do you think?

**Comment by imballinst** [comment]
Oh! But, if what you meant as we use the footnote syntax, I think we can definitely do that (so the footnote syntax becomes a button). Let me try it tomorrow.

**Comment by dio** [comment]
> Oh! But, if what you meant as we use the footnote syntax, I think we can definitely do that (so the footnote syntax becomes a button). Let me try it tomorrow.

Yeap. Exactly this.

**Comment by imballinst** [comment]
@dio WDYT about this? The implementation is with popper.js -- I used the one for rather imperative one, `createPopper`, instead of `usePopper`, because the HTML structures are created in the plugin instead of in React: https://imballinst.github.io/protosaurus/pr-32/docs/test/

Some are still unstyled:

https://user-images.githubusercontent.com/7077157/155992346-3c37505a-3f0a-4eaf-af04-78d5434b41b9.mov

**Comment by dio** [comment]
> @dio WDYT about this? The implementation is with popper.js -- I used the one for rather imperative one, `createPopper`, instead of `usePopper`, because the HTML structures are created in the plugin instead of in React: [imballinst.github.io/protosaurus/pr-32/docs/test](https://imballinst.github.io/protosaurus/pr-32/docs/test/)
> 
> Some are still unstyled:

Nice. This is perfect. Thank you!

**Comment by imballinst** [comment]
Hum, let me see. There are 2 other use cases that this PR has yet to address:

1. Normal footnotes (outside of code block) stops working
2. Footnotes inside a comment in the code block isn't possible yet

```
Test [^1]
```

````
```protosaurus
message Test {
  // Test comment [^1]
}
```
````

Both of above don't work yet. Let me see what I can do.

**Comment by imballinst** [comment]
@dio new updates:

(images are outdated, please see the images in the next comment instead)

1. **Annotations now replace footnotes effectively**: paragraph, protosaurus codeblock comment, and protosaurus codeblock code + comment.
2. **Improve icon button color:** I use "info" color button, but feedback are open with regards to its accessibility
3. **Improve styling**: add box shadow for better "focus" in light mode and use margin and trim left/right spaces when it comes to being in the middle of text with spaces)

Let me know what you think!

**Comment by imballinst** [comment]
Iuh, actually, I changed the icon. Instead of going for the solid info icon, I went for the outlined one. I think it's easier to recognize it than the solid one.

image

image

File: packages/docusaurus-preset/src/index.ts [hunk 1]
```diff
@@ -29,11 +30,15 @@ export default function protosaurusPreset(
 ): Preset {
   const presetOpts = options;
   const rehypePlugin = [protoMessageRehypePlugin, { siteDir: context.siteDir }];
+  const remarkPlugin = [remarkPluginAnnotations, { siteDir: context.siteDir }];
 
   if (typeof presetOpts?.docs === 'object') {
     presetOpts.docs.rehypePlugins = presetOpts?.docs.rehypePlugins
       ? [...presetOpts?.docs.rehypePlugins, rehypePlugin]
       : ([rehypePlugin] as any);
+    presetOpts.docs.remarkPlugins = presetOpts?.docs.remarkPlugins
```
Comment (dio) [inline packages/docusaurus-preset/src/index.ts:39]: A question, when we have another remark plugin, can we add it from the docusaurus config? I mean, can this preset accept args? Is this presetOpts the args?

Comment (dio) [reply packages/docusaurus-preset/src/index.ts:39]: NVM seems like so. This is cool. Thanks!

**Review by dio — APPROVED** [review]

File: packages/docusaurus-preset/src/index.ts (hunk 1 above)
Comment (imballinst) [reply packages/docusaurus-preset/src/index.ts:39]: Yes, if we want to pass another options to override our default config, perhaps we can pass it from `docusaurus.config.js`, something like:

```js
{
  docs: {}, // docusaurus stuff
  protosaurus: {} // our stuff
}
```

Then, we can access it with `presetOptions.protosaurus`.

**Pull request merged**